### PR TITLE
Rewrite SelfDocumentID as a property wrapper

### PR DESCRIPTION
**Pull request by wilhuff**
Make wrapped types conform to DocumentReferenceConvertible and use that 
protocol to implement the conversion.

### CODE REVIEW

File: Firestore/Swift/Source/Codable/DocumentID.swift [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+ * limitations under the License.
+ */
+
+import FirebaseFirestore
+
+/// A type that can convert to itself from a Firestore `DocumentReference`,
+/// which makes it suitable for use with the `@DocumentID` property wrapper.
+///
+/// Firestore includes extensions that make `String` and `DocumentReference`
+/// conform to `@DocumentReferenceConvertible`.
```
Comment (wu-hui) [inline Firestore/Swift/Source/Codable/DocumentID.swift]: No `@` for protocol.

File: Firestore/Swift/Source/Codable/DocumentID.swift [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+///
+/// Note that Firestore ignores fields annotated with `@DocumentID` when writing
+/// so there is no requirement to convert from the wrapped type back to a
+/// `DocumentReference`. In particular, this means that conversions can be
+/// lossy without any problem.
```
Comment (wu-hui) [inline Firestore/Swift/Source/Codable/DocumentID.swift]: It took me about 5 seconds to understand what `lossy` means here, maybe we can make it more straightforward? I don't have good suggestions though.

File: Firestore/Swift/Source/Codable/DocumentID.swift [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+extension DocumentReference: DocumentReferenceConvertible {
+  public static func convert(from documentReference: DocumentReference) -> Self {
+    // Swift complains that values of type DocumentReference cannot be returned
+    // as Self which is nonsensical. The cast forces this to work.
+    return documentReference as! Self
+  }
+}
+
+/// An internal protocol that allows FirestoreDecoder to test if a type is a
```
Comment (wu-hui) [inline Firestore/Swift/Source/Codable/DocumentID.swift]: `Firestore.Decoder`?

**Comment by wilhuff** [comment]
Addressed your feedback in the PR plus the outcome of API discussion last week.

**Pull request merged**